File: python-backend/demo_runtime.py
"""Opt-in demo isolation, logical time, and MQTT run fencing.

The control document lives in the explicitly local demo database. Each run owns
its own collection prefix, so even a delayed operation cannot write a new run.
Wall-clock timestamps remain available for retention and connection health.
"""
from __future__ import annotations

import asyncio
import contextvars
import json
import os
import re
import time
import uuid
from contextlib import asynccontextmanager
from datetime import datetime
from urllib.parse import urlparse

CONTROL_COLLECTION = "_demoControl"
WORKER_COLLECTION = "_demoWorkers"
CURRENT = contextvars.ContextVar("harboursense_demo_context", default=None)
WRITE_METHODS = {"insert_one", "insert_many", "update_one", "update_many",
                 "replace_one", "delete_one", "delete_many", "find_one_and_update",
                 "find_one_and_replace", "find_one_and_delete", "bulk_write"}


class DemoRunEnded(RuntimeError):
    pass


def demo_enabled(env=None):
    return (os.environ if env is None else env).get("DEMO_MODE", "").lower() == "true"


def validate_demo_settings(settings, env=None):
    env = os.environ if env is None else env
    if not demo_enabled(env):
        raise ValueError("DEMO_MODE=true is required")
    uri = urlparse(settings.uri)
    name = settings.database_name
    if (uri.scheme != "mongodb" or uri.hostname not in {"localhost", "127.0.0.1", "::1", "mongo", "mongo-demo"}
            or uri.username or uri.password or "," in uri.netloc
            or not re.fullmatch(r"harboursense_demo(?:_test_[a-z0-9_]+)?", name)
            or uri.path.strip("/") not in {"", name}):
        raise ValueError("Demo requires an unauthenticated local MongoDB URI and an owned harboursense_demo database")
    if env.get("MQTT_MODE", "local") != "local" or env.get("MQTT_BROKER_HOST", "localhost") not in {"localhost", "127.0.0.1", "mqtt", "mqtt-demo"}:
        raise ValueError("Demo requires the local MQTT broker")


def logical_ms(state, wall_ms=None):
    clock = state.get("clock", {})
    value = float(clock.get("baseMs", 0))
    if state.get("status") == "running":
        value += max(0, (time.time() * 1000 if wall_ms is None else wall_ms) - clock.get("anchorWallMs", 0)) * state.get("speed", 1)
    return value


def run_prefix(run_id):
    if not re.fullmatch(r"[a-f0-9]{32}", run_id):
        raise ValueError("Invalid demo run identity")
    return f"run_{run_id}_"


class RunCollection:
    def __init__(self, raw, context):
        self.raw, self.context = raw, context

    def __getattr__(self, name):
        method = getattr(self.raw, name)
        if name not in WRITE_METHODS:
            return method

        async def write(*args, **kwargs):
            async with self.context.operation():
                return await method(*args, **kwargs)
        return write


class RunDatabase:
    def __init__(self, base, context, guarded=True):
        self.base, self.context, self.guarded = base, context, guarded
        self.name = base.name
        self.client = base.client

    def __getitem__(self, name):
        raw = self.base[run_prefix(self.context.run_id) + name]
        return RunCollection(raw, self.context) if self.guarded else raw

    def __getattr__(self, name):
        if name.startswith("_"):
            raise AttributeError(name)
        return self[name]

    async def command(self, *args, **kwargs):
        return await self.base.command(*args, **kwargs)

    async def list_collection_names(self):
        prefix = run_prefix(self.context.run_id)
        return [name[len(prefix):] for name in await self.base.list_collection_names() if name.startswith(prefix)]


class DemoContext:
    def __init__(self, base, run_id, role="api"):
        self.base, self.run_id, self.role = base, run_id, role
        self.state = {}
        self.ready = False
        self.mqtt_connected = False
        self.db = RunDatabase(base, self)

    async def refresh(self):
        state = await self.base[CONTROL_COLLECTION].find_one({"_id": "active"})
        if not state or state.get("runId") != self.run_id:
            raise DemoRunEnded("Demo run was replaced")
        self.state = state
        return state

    async def checkpoint(self):
        while True:
            state = await self.refresh()
            if state["status"] == "running":
                return state
            if state["status"] == "resetting":
                raise DemoRunEnded("Demo run stopped")
            await asyncio.sleep(.05)

    @asynccontextmanager
    async def operation(self):
        while True:
            await self.checkpoint()
            result = await self.base[CONTROL_COLLECTION].update_one(
                {"_id": "active", "runId": self.run_id, "status": "running"},
                {"$inc": {"activeOperations": 1}})
            if result.modified_count:
                break
        try:
            yield
        finally:
            await self.base[CONTROL_COLLECTION].update_one(
                {"_id": "active", "runId": self.run_id}, {"$inc": {"activeOperations": -1}})

    async def sleep(self, seconds):
        state = await self.checkpoint()
        target = logical_ms(state) + seconds * 1000
        while logical_ms(await self.checkpoint()) < target:
            await asyncio.sleep(.025)

    async def heartbeat(self, status="ready", **extra):
        self.ready = status == "ready"
        if "mqttConnected" in extra:
            self.mqtt_connected = extra["mqttConnected"]
        await self.base[WORKER_COLLECTION].update_one(
            {"_id": f"{self.run_id}:{self.role}"}, {"$set": {"role": self.role, "runId": self.run_id, "status": status,
              "updatedAt": datetime.utcnow(), "mqttConnected": self.mqtt_connected, **extra}}, upsert=True)


async def simulation_sleep(seconds):
    context = CURRENT.get()
    if context:
        await context.sleep(seconds)
    else:
        await asyncio.sleep(seconds)


def simulation_time():
    context = CURRENT.get()
    if context and context.state:
        return context.state.get("startedWallMs", 0) / 1000 + logical_ms(context.state) / 1000
    return time.time()


def simulation_datetime():
    return datetime.utcfromtimestamp(simulation_time()) if CURRENT.get() else datetime.now()


async def simulation_checkpoint():
    context = CURRENT.get()
    if context:
        await context.checkpoint()


class DemoMqttClient:
    def __init__(self, client, context):
        self.client, self.context = client, context

    def __getattr__(self, name):
        return getattr(self.client, name)

    async def subscribe(self, topic, *args, **kwargs):
        kwargs.setdefault("qos", 1)
        return await self.client.subscribe(topic, *args, **kwargs)

    async def publish(self, topic, payload, *args, **kwargs):
        data = json.loads(payload) if isinstance(payload, (str, bytes, bytearray)) else dict(payload)
        if data.get("runId", self.context.run_id) != self.context.run_id:
            raise DemoRunEnded("Refusing a previous-run publication")
        async with self.context.operation():
            state = await self.context.base[CONTROL_COLLECTION].find_one_and_update(
                {"_id": "active", "runId": self.context.run_id}, {"$inc": {"eventSequence": 1}}, return_document=True)
            data.update(sequence=state["eventSequence"], runId=self.context.run_id, scenarioId=state["scenarioId"],
                        eventId=data.get("eventId", uuid.uuid4().hex),
                        simulatedTimeMs=logical_ms(state), wallTime=datetime.utcnow().isoformat() + "Z")
            kwargs.setdefault("qos", 1)
            return await self.client.publish(topic, json.dumps(data, default=str), *args, **kwargs)


async def accept_demo_event(payload):
    context = CURRENT.get()
    if not context:
        return True
    if not isinstance(payload, dict) or payload.get("runId") != context.run_id or not payload.get("eventId"):
        return False
    await context.checkpoint()
    return not await context.db.demoInbox.find_one({"_id": payload["eventId"]})


async def mark_demo_event(payload):
    context = CURRENT.get()
    if context and payload.get("eventId"):
        await context.db.demoInbox.update_one({"_id": payload["eventId"]},
            {"$set": {"processedAt": datetime.utcnow()}}, upsert=True)


async def run_worker(role, base, callback):
    """Rebind a whole worker after reset; cancellation releases old operations."""
    while True:
        task = None
        context = None
        try:
            state = await base[CONTROL_COLLECTION].find_one({"_id": "active"})
            if not state:
                await asyncio.sleep(.25)
                continue
            context = DemoContext(base, state["runId"], role)
            context.state = state
            token = CURRENT.set(context)
            task = asyncio.create_task(callback(context))
            CURRENT.reset(token)
            while True:
                current = await base[CONTROL_COLLECTION].find_one({"_id": "active"})
                if not current or current["runId"] != context.run_id:
                    break
                if task.done():
                    error = task.exception()
                    if error and not isinstance(error, DemoRunEnded):
                        raise error
                    if current["status"] not in {"complete", "resetting", "failed"}:
                        break
                await context.heartbeat(("ready" if context.ready else "starting") if not task.done() else "stopped")
                await asyncio.sleep(.5)
        except asyncio.CancelledError:
            raise
        except Exception as exc:
            if context:
                try:
                    await context.heartbeat('starting', mqttConnected=False, error=type(exc).__name__)
                except Exception:
                    pass
            await asyncio.sleep(1)
        finally:
            if task:
                task.cancel()
                await asyncio.gather(task, return_exceptions=True)

File: port-sim/lib/demo-runtime.js
/* Shared demo clock and collection fencing; normal live execution is untouched. */
const { AsyncLocalStorage } = require('async_hooks');
const crypto = require('crypto');
const storage = new AsyncLocalStorage();
const WRITES = new Set(['insertOne', 'insertMany', 'updateOne', 'updateMany', 'replaceOne',
  'deleteOne', 'deleteMany', 'findOneAndUpdate', 'findOneAndReplace', 'findOneAndDelete', 'bulkWrite']);
const wallSleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

class DemoRunEnded extends Error {}

function validateDemoSettings(settings, env = process.env) {
  const uri = new URL(settings.uri);
  if (env.DEMO_MODE !== 'true' || uri.protocol !== 'mongodb:'
      || !['localhost', '127.0.0.1', '[::1]', 'mongo', 'mongo-demo'].includes(uri.hostname)
      || uri.username || uri.password || uri.host.includes(',')
      || !/^harboursense_demo(?:_test_[a-z0-9_]+)?$/.test(settings.databaseName)
      || (uri.pathname !== '/' && uri.pathname !== '' && uri.pathname !== `/${settings.databaseName}`)
      || (env.MQTT_MODE || 'local') !== 'local'
      || !['localhost', '127.0.0.1', 'mqtt', 'mqtt-demo'].includes(env.MQTT_BROKER_HOST || 'localhost')) {
    throw new Error('Demo requires the owned local MongoDB database and local MQTT broker');
  }
}

function logicalMs(state, wall = Date.now()) {
  return (state.clock?.baseMs || 0) + (state.status === 'running'
    ? Math.max(0, wall - state.clock.anchorWallMs) * state.speed : 0);
}

function runPrefix(runId) {
  if (!/^[a-f0-9]{32}$/.test(runId)) throw new Error('Invalid demo run identity');
  return `run_${runId}_`;
}

class DemoContext {
  constructor(base, state, role) {
    this.base = base;
    this.state = state;
    this.runId = state.runId;
    this.role = role;
    this.stopped = false;
    this.mqtt = null;
    this.db = { collection: (name) => this.collection(name) };
  }

  async refresh() {
    if (this.stopped) throw new DemoRunEnded('Demo worker stopped');
    const state = await this.base.collection('_demoControl').findOne({ _id: 'active' });
    if (!state || state.runId !== this.runId) throw new DemoRunEnded('Demo run was replaced');
    this.state = state;
    return state;
  }

  async checkpoint() {
    while (true) {
      const state = await this.refresh();
      if (state.status === 'running') return state;
      if (state.status === 'resetting') throw new DemoRunEnded('Demo run ended');
      await wallSleep(25);
    }
  }

  async operation(callback) {
    while (true) {
      await this.checkpoint();
      const result = await this.base.collection('_demoControl').updateOne(
        { _id: 'active', runId: this.runId, status: 'running' }, { $inc: { activeOperations: 1 } });
      if (result.modifiedCount) break;
    }
    try { return await callback(); }
    finally {
      await this.base.collection('_demoControl').updateOne(
        { _id: 'active', runId: this.runId }, { $inc: { activeOperations: -1 } });
    }
  }

  collection(name) {
    const raw = this.base.collection(runPrefix(this.runId) + name);
    return new Proxy(raw, { get: (target, key) => {
      const member = target[key];
      if (typeof member !== 'function') return member;
      if (!WRITES.has(key)) return member.bind(target);
      return (...args) => this.operation(() => {
        if (name === 'edgeRuntime' && ['updateOne', 'updateMany', 'findOneAndUpdate'].includes(key) && args[1]?.$set) {
          args[1] = { ...args[1], $set: { ...args[1].$set, demoUpdatedSimMs: logicalMs(this.state) } };
        }
        return member.apply(target, args);
      });
    } });
  }

  async sleep(ms) {
    const target = logicalMs(await this.checkpoint()) + ms;
    while (logicalMs(await this.checkpoint()) < target) await wallSleep(20);
  }

  async heartbeat(status = 'ready', error = null) {
    await this.base.collection('_demoWorkers').updateOne({ _id: `${this.runId}:${this.role}` }, { $set: {
      role: this.role, runId: this.runId, status, updatedAt: new Date(), mqttConnected: Boolean(this.mqtt?.connected), error,
    } }, { upsert: true });
  }

  async publish(topic, payload, callback) {
    try {
      const data = typeof payload === 'string' || Buffer.isBuffer(payload) ? JSON.parse(payload) : { ...payload };
      if (data.runId && data.runId !== this.runId) throw new DemoRunEnded('Previous-run publication rejected');
      const state = await this.checkpoint();
      const sequenceResult = await this.base.collection('_demoControl').findOneAndUpdate(
        { _id: 'active', runId: this.runId }, { $inc: { eventSequence: 1 } }, { returnDocument: 'after' });
      const sequenceState = sequenceResult?.value || sequenceResult;
      Object.assign(data, { sequence: sequenceState?.eventSequence, runId: this.runId, scenarioId: state.scenarioId,
        eventId: data.eventId || crypto.randomUUID(), simulatedTimeMs: logicalMs(state), wallTime: new Date().toISOString() });
      await this.db.collection('mqttOutbox').updateOne({ _id: data.eventId }, {
        $setOnInsert: { topic, payload: data, producer: this.role, delivered: false, createdAt: new Date() },
      }, { upsert: true });
      if (callback) callback(null);
    } catch (error) {
      if (callback) callback(error);
      else throw error;
    }
  }

  async flushOutbox() {
    while (true) {
      await this.checkpoint();
      if (this.mqtt?.connected) {
        const pending = await this.db.collection('mqttOutbox').find({ delivered: false, producer: this.role }).sort('createdAt', 1).limit(100).toArray();
        for (const item of pending) {
          await this.checkpoint();
          await new Promise((resolve, reject) => this.mqtt.publish(item.topic, JSON.stringify(item.payload), { qos: 1 }, (error) => error ? reject(error) : resolve()));
          await this.db.collection('mqttOutbox').updateOne({ _id: item._id }, { $set: { delivered: true } });
        }
      }
      await wallSleep(30);
    }
  }

  async accept(payload) {
    if (payload?.runId !== this.runId || !payload.eventId) return false;
    await this.checkpoint();
    return !await this.db.collection('mqttInbox').findOne({ _id: payload.eventId });
  }

  async acknowledge(payload) {
    await this.db.collection('mqttInbox').updateOne({ _id: payload.eventId }, { $set: { processedAt: new Date() } }, { upsert: true });
  }
}

function simulationNow() {
  const context = storage.getStore();
  return context ? context.state.startedWallMs + logicalMs(context.state) : Date.now();
}
function simulationSleep(ms) {
  const context = storage.getStore();
  return context ? context.sleep(ms) : wallSleep(ms);
}
function currentContext() { return storage.getStore(); }

module.exports = { DemoContext, DemoRunEnded, validateDemoSettings, logicalMs, runPrefix,
  simulationNow, simulationSleep, currentContext, storage, wallSleep };
